### ISSUE TITLE
testing

### ISSUE THREAD

**Issue by PavelVozenilek** [c3lang/c3c#11]
There's one project taking testing seriously, SQLite:
https://www.sqlite.org/testing.html

My ideal is to make testing trivial.

* It is possible to implement it in C++:
    https://www.codeproject.com/Articles/9757/Minimal-unit-test-library-for-C
* It is possible to do it in C, with just one non-standard trick (`__attribute__((constructor))`, another trick exists in VC++).
* It is possible to do it in Nim:
    https://forum.nim-lang.org/t/653

---------------

How the tests should look like:

```
...
... // code
...

TEST()
{
  assert(1 + 1 == 2);
}

TEST()
{
  assert(2 + 2 == 5);
}

...
... // more code
...
```

That's all. Not a single character more. No unique names, no manual registration of tests, no `#include`s of frameworks. No _set up_/_tear down_ vomit.  No other infrastructure.

Just type these 8 characters and 3 newlines and have new empty test ready to run. I can do it in second or two. Anything taking more time is unacceptable.

----------------

It should be possible to run tests both in debug or in release mode. The latter is important, to make sure the optimizer or I didn't screw up.

I envision copying section for `debug`/`release` target under name `debug-with-tests`/`release-with-tests`, and adding `TESTS` define there. That should be all to enable the testing.

--------------

The old `assert` is good enough to do the tests. I'd seen test frameworks defining dozen assert like functions, e.g. specific assert to compare two floating point numbers. This is not needed.

However, the tests should be run also in release mode. `assert`s usually works only in debug mode, because there are so many of them. They would kill release mode performance. I am defensive programmer, I use `assert` more than any other construct. I check even completely trivial things, and found at least one bug in compiler this way.

So my proposal is to have **3 variants** of asserts:

1. the traditional `assert`. Not active in release modes. Could be placed anywhere.

2. "assert for release mode". Active if and only if tests are enabled. Available _only_ inside `TEST()`, nowhere else. I named it `verify`.

    ```
    TEST()
    {
      assert(1 + 1 == 2); // does the check in debug mode, in release it is empty test
    }
    
    TEST()
    {
      verify(1 + 1 == 2); // works both in debug and release modes. 
                          // MUST NOT be used outside tests
    } 
    ```
3. "assert used for impossible situations". Many error situations are almost impossible. I place `assert(false)` into such paths, to be doubly sure.

    It would look like this:
    ```
    ...
    if (impossible-situation) {
       assert(false); // cannot happen
       return;
    }
    ...
    ```

  However, within tests I try to provoke such impossible situations, and then check whether the code reacts properly. The old `assert(false)` would always fire, which I do not want, in this very specific situation.

  I solved it by adding yet another form of assert unimaginatively named `assert2`. It fires only in debug mode and only when _no_ test is currently running. In other words, it does not fire during the actual testing.

   It would look like this:

    ```
    bool foo() {
       ...
       if (impossible-situation) {
         assert2(false);
         return false;
       }
       return true;
    }

    TEST()
    {
       ... // prepare impossible situation
      verify(foo() == false); // the assert2 won't fire, the test would pass
    }

    ... // normal code
    foo(); // if impossible situation happens, assert2 fires.
    ```

All kind of asserts should be able to find out, whether there's a test currently running, and if so, then in what is the file/line of the test. This information should then be shown. 

---------------

Test runner. This is how I call the code actually running the tests.

The most basic test runner could be provided in stdlib. Just two functions is all needed.

It could look like:

```
TEST()
{
}

void main()
{
  #ifdef TESTS
    if (some hint-indicating-that-I-want-to-run-tests) 
    {
      run_all_tests(); // stdlib function which magically invokes all tests
    }
  #endif
 
   ... // normal code
}
```

Or I may like to run tests only from recently modified source files. That's surprisingly easy to do in C.

How it may look like:

```
void main()
{
  #ifdef TESTS
    // stdlib function which magically invokes tests
    // from files modified within last 2 hours.
    //
    // Run *every* time the application starts 
    // since it would cause only minimal delay.
    run_recent_tests(); 
  #endif
 
   ... // normal code
}
```

-----------------

Developer may develop his own test runners. For this tests could have attributes. Attributes should have completely free format.

It may look like:

```
TEST() // no attributes test
{
}

TEST(name = "aaa") // named test
{
}

TEST( time < 1 ms) // test which should run under 1 millisecond
{
}

// test with many attributes
TEST(name = "aaa" | foo | bar(baz) |  10 ms < time < 30 ms | something^2 )
{
}
```
"`|`" is attribute separator.

Custom test runner would be able to access the attributes, and act accordingly. 

For example, if there's explicit time limit, the runner would check that the test is faster. If it not, it could repeat the test up to 5x, and if it is always slower, show error. (Test could have some reasonable default timeout, to catch unexpectedly slow tests. I used 30 milliseconds. Whatever was slower had to be annotated.)

-----------------

**Tests could serve as leak detectors!**

Test runners could also check other things. For example, make sure that the test does not leak. It was quite hard to implement, but it did miracles for me. No more silly leaks, tests caught everything. These checks were done automatically, no extra work for me, no silly noise within the test.

---------------

One test attribute needs compiler support. Sometimes I want to be sure that certain code _does not_ compile. Uncompilable test could do it:

```
TEST(uncompilable) // no other attributtes allowed together with "uncompilable"
{
  ... // something what compiler refuses
}
```

The compiler would verify that the code fails. The test itself wont be passed to any test runner.

Content of uncompilable test should be syntactically correct. E.g. no misplaced parenthesis. If such tests are needed (I have doubts), then the uncompilable feature could be extended.

--------------

These test could be also used to collect profiling data.

```
TEST() // ordinary test
{
}

TEST(profiling)
{
  ... // do something useful for profiling
}
```
Special test runner would pick up only the tests intended for profiling and run them. This way relevant data could be gathered by the profiles.

Performance regression data could collected similarly.

-------------

Here's a biggie. It is often desirable to run tests checking code which deals with files, network, databases and other slow and clumsy things. However, I do not want these tests to take forever.

What I want is to replace slow APIs like `fopen` with something on mine, something what only mocks the real functionality, in no time. And I want to do it easily, with no complicated mocking etc.

How it may look like:

```
TEST()
{
   replace fopen = func () { ... } // replaced stldib fopen with my code
   replace flose = func () { ... } // my fclose alternative
    ...
    ...  // code indirectly invoking many fopen/fclose
}
```

The compiler would understand what I am trying to do. It would (if testing is allowed) replace all calls to `fopen` to invoking function pointer, which by default points to the old good `fopen`.

However, inside that test, `fopen` pointer would be replaced by my own function. At the end of the test, the compiler would return back the original `fopen`.

Similarly, constants could be replaced by value in memory, set and reset within relevant tests. E.g I may want to change some timeout constant to do the test quickly.

```
TEST()
{
  replace TCPIP_TIMEOUT = 0.001; // default is 2 hours
  ...
  ... code checking whether timeout works
}
```

-----------------

Handy feature from Zig language. Sometimes tests share lot of initial and ending code. This may be solved like this:

```
TEST()
{
  common-set-up-code();

  TEST()
  {
    test1-code();
  }

  TEST()
  {
    test2-code();
  }

  common-tear-down-code()
}
```

This would transform into:

```
TEST()
{
  common-set-up-code();
  test1-code();
  common-tear-down-code()
}

TEST()
{
  common-set-up-code();
  test2-code();
  common-tear-down-code()
}
```
-----------

Handy idea from D language. Tests do not need to be only at top level.

```
struct x
{
  ...
  TEST()
  {
    .. test related to this structure
  }
}

void foo()
{
  if (impossible) {
    TEST()
    {
      ... // test invoking the impossible situation, makes sense to be here
    }
    return;
  }
}
```

These tests are nice to have feature for special circumstances.

-------------

Tests should have access to all parts of structure, whether private or public. Really thorough testing needs this.

------------

Having many tests makes source file potentially very, very long. (I had no problem to have  thousands of lines of code with tests in a single file.) I propose to have "test companion file" feature.

If `something.x` is source file, then `something.tests.x` would be the tests companion file. This companion file would be allowed to contain only tests and their helpers. Nothing could be exported out of this file. Tests in such file would have full access to the "parent" `something.x`, as if it was appended to the end of this parent file.

Potentially useful sub-feature is to give reader hints about existence of tests in companion file:

```
--- something.x = source file ---
...
TEST() // ordinary test
{
}
// forward of a test, it has to be in test companion under this name
TEST(name = "x"); 
TEST(name = "y") // another forward

TEST() // normal test again
{
}

---- something.tests.x  = test companion file ---

// all forward tests need to be here, in the same order, 
// for easier lookup
TEST(name = "x")
{
}

TEST(name = "y")
{
}
```

Both options to place tests, in source file and in test companion file should be supported at the same time. Sometimes the first way is better, sometimes the other one.

---------

The testing framework expects that individual test either finishes (= everything works as expected), or the execution ends with an `assert` being fired. This stops the program (e.g. breaking into the debugger).

I do not support failing tests. If something goes wrong and assert fires, that bug should be fixed. Period. Collecting failures makes no sense.

However, if failing tests are desirable, they could be supported. It would increase complexity of the testing infrastructure, though.

---------------

I hope I didn't kill anyone with this long text. In D language forum someone once claimed that D's `unittest` is the most useful feature of the language.

It is possible to do even better, and it is not that hard, compared to many language features. Large parts of test support could be done via libraries.

**Comment by lerno** [comment]
Let me address this one pieces at a time. It will be simple to generate a macro that adds all functions marked with the attribute @test (or similar) to an array at compile time. Using this it’s extremely easy to create a unittest framework.

**Comment by PavelVozenilek** [comment]
I see problems with @test annotation:

1. It is not minimal. Too many characters to type. Also not visually distinctive enough. 
2. You would need to invent a name for this function. This is pain, and useless. I just want to make a test here and now, without any fuss. Think of project with tens of thousand tests. Someone may call a test function. IDE may show them. Name clashes.


I am quite obsessed with testing, and how to remove noise around it.

* Zig almost fits my ideal, except for requiring useless string parameter for every test.  Also no way to use my own test runner, so far.
* D is almost ideal, except that you need to run all tests all the time (one could reimplement test runner with [breathtakingly complicated](https://forum.dlang.org/post/vblsrdsqbaxthuomikqa@forum.dlang.org) metaprogramming tricks). 
* Nim allows to do everything with its metaprogramming, but this also means compiler is not aware of tests, which (probably, I am not expert here) limits possible advanced checks.
* The worst language in this regard is Pony. They had show an example of test, and I counted 1 assert, 17 helper lines with a need to create new class. It also required manual registration.

**Comment by lerno** [comment]
Using compile time reflection, you could also say "run everything starting with test as test". Or "run everything in this directory as test".

Even simpler would actually be a weird idea I call a "run include"

```
@run_include("foo.sh", $some_param, "-x", $another_param);
```

Basically you include the stdout from a program that is run during compilation.

**Comment by PavelVozenilek** [comment]
Reply to https://github.com/c3lang/c3c/issues/11#issuecomment-521938236:

Even if everything testable starts with `test`, you still need to invent the remaining part of the name. Plus the need to type more than absolutely necessary.

Having separate directory for tests is bad idea. Test should be as close to tested code as only possible. Right below the function (or if they get too big in test companion file, with forward reference in the main file).

I once tried to place them all at the end of source file. It was disaster, I couldn't orient among them, had no idea if some feature is tested or not.

------------

I do not understand the "run include" idea.

**Comment by lerno** [comment]
I used to think that tests should be together with the source code, but have since changed my opinion:

1. It increases the length of reach file.
2. While functions are constantly refactored, tests should not be and might actually be rather ugly as long as they correctly test things. The discrepancy between the two styles is a problem.
3. I find it harder to distinguish working code from its tests. Often tests many times longer than the code they're testing. Having those thrown in between rows code makes code reading harder.
4. It should be easy to compile just the code without the tests.
5. To move something from a library into the language should in my opinion only be done if it's not possible to express it in the language itself. I don't feel that is the case here.

**Comment by PavelVozenilek** [comment]
Reply to https://github.com/c3lang/c3c/issues/11#issuecomment-521989937:

ad 1) Right, heavy testing expands source file size considerably. For this reason I wish to have the "test companion file" feature, dedicated test file sitting just next to the "main" source file. This proximity (in filesystem location and in file name) gives some chance to discover relevant tests. Moving the things into some `test/` directory would make them virtually disappear.

Hypothetical, very advanced IDE would be able to hide the tests.

-----------
ad 2) Ugliness of tests: test companion files are intended to keep such tests out of sight.

-----------

ad 3) Hard to distinguish between the code and tests: that's why I picked syntax:

```
TEST()
{
   ...
}
```

I found this to be good visual clue to spot the tests. That's also why I am against tests being "similar" to ordinary functions.

----------------
ad 3a) Too big tests should be placed into test companion file.


-----------

ad 4) Ability to switch tests/no-tests: that's the task of `TESTS` define. One change in project configuration file, and tests are enabled/disabled.

-----------

ad 5) Yes. Most of the functionality could be implemented by a library. However, for following features compiler support is needed:

* Tests should be able to inspect everything, even private parts of data structures. 
* Compiler should wrap tests into implicit `#ifdef TESTS ... #endif` brackets (or some equivalent). Typing these by hand is nuisance.
* Ability to pass free format attributes to the tests. Without attributes testing subsystem would be quite limited. Using not-a-free format (e.g. `key=value` pairs) is possible, but clumsy.
* Doing leak detection may or may not require help by the compiler. In C it was not needed, because the language allows to hijack allocations completely.
* The **big** feature, ability to replace (to mock) specified functions during test duration requires compiler support. The implementation is easy, but cannot be done by library code.
* Nested tests (feature inspired by Zig) seem to require compiler support.
* Ability to place test inside structures or within some code (as in D) would require compiler support.
* Test companion files would need compiler support: full access to "parent" source, ensuring that the companion file doesn't export anything, verifying that all forward declarations of tests exist.


Only a language with extremely advanced metaprogramming capabilities could implement some or all of these features by a library. (For example, I once got confirmation that in Nim one _could_ implement tests inside function body.)
